pkg/storage/imagestore/imagestore.go

```diff
@@ -0,0 +1,2253 @@
+
+			return []byte{}, zerr.ErrRepoNotFound
+		}
+
+		is.log.Error().Err(err).Str("dir", dir).Msg("failed to read index.json")
+
+		return []byte{}, err
+	}
+
+	return buf, nil
+}
+
+// DeleteBlob removes the blob from the repository.
+func (is *ImageStore) DeleteBlob(repo string, digest godigest.Digest) error {
+	var lockLatency time.Time
+
+	if err := digest.Validate(); err != nil {
+		return err
+	}
+
+	is.Lock(&lockLatency)
+	defer is.Unlock(&lockLatency)
+
+	return is.deleteBlob(repo, digest)
+}
+
+func (is *ImageStore) deleteBlob(repo string, digest godigest.Digest) error {
+	blobPath := is.BlobPath(repo, digest)
+
+	_, err := is.storeDriver.Stat(blobPath)
+	if err != nil {
+		is.log.Error().Err(err).Str("blob", blobPath).Msg("failed to stat blob")
+
+		return zerr.ErrBlobNotFound
+	}
+
+	// first check if this blob is not currently in use
+	if ok, _ := common.IsBlobReferenced(is, repo, digest, is.log); ok {
+		return zerr.ErrBlobReferenced
+	}
+
+	if fmt.Sprintf("%v", is.cache) != fmt.Sprintf("%v", nil) {
+		dstRecord, err := is.cache.GetBlob(digest)
+		if err != nil && !errors.Is(err, zerr.ErrCacheMiss) {
+			is.log.Error().Err(err).Str("blobPath", dstRecord).Msg("dedupe: unable to lookup blob record")
+
+			return err
+		}
+
+		// remove cache entry and move blob contents to the next candidate if there is any
+		if ok := is.cache.HasBlob(digest, blobPath); ok {
+			if err := is.cache.DeleteBlob(digest, blobPath); err != nil {
+				is.log.Error().Err(err).Str("digest", digest.String()).Str("blobPath", blobPath).
+					Msg("unable to remove blob path from cache")
+
+				return err
+			}
+		}
+
+		// if the deleted blob is one with content
+		if dstRecord == blobPath {
+			// get next candidate
+			dstRecord, err := is.cache.GetBlob(digest)
+			if err != nil && !errors.Is(err, zerr.ErrCacheMiss) {
+				is.log.Error().Err(err).Str("blobPath", dstRecord).Msg("dedupe: unable to lookup blob record")
+
+				return err
+			}
+
+			// if we have a new candidate move the blob content to it
+			if dstRecord != "" {
+				/* check to see if we need to move the content from original blob to duplicate one
+				(in case of filesystem, this should not be needed */
+				binfo, err := is.storeDriver.Stat(dstRecord)
+				if err != nil {
+					is.log.Error().Err(err).Str("path", blobPath).Msg("rebuild dedupe: failed to stat blob")
+
+					return err
+				}
+
+				if binfo.Size() == 0 {
+					if err := is.storeDriver.Move(blobPath, dstRecord); err != nil {
+						is.log.Error().Err(err).Str("blobPath", blobPath).Msg("unable to remove blob path")
+
+						return err
+					}
+				}
+
+				return nil
+			}
+		}
+	}
+
+	if err := is.storeDriver.Delete(blobPath); err != nil {
+		is.log.Error().Err(err).Str("blobPath", blobPath).Msg("unable to remove blob path")
+
+		return err
+	}
+
+	return nil
+}
+
+func (is *ImageStore) garbageCollectIndexReferrers(repo string, index ispec.Index) (bool, error) {
```

Comment:
Can we move this function just after the declaration of garbageCollect? It's hard to follow the code as the functions are not called in the order of which they are defined in the file.